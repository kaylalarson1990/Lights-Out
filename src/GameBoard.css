.header {
  display: flex;
  flex-direction: column;
  margin: 0 4em;
}

.restart {
  background-color: #78FFD6;
  border: 1px #1a1111 solid;
  border-radius: 1em;
  color: #1a1111;
  font-weight: 700;
  padding: 0.6em;
  width: 8em;
}

.title {
  color: #1a1111;
  font-weight: 300;
}

.game-options {
  display: flex;
  color: #1a1111;
  margin-bottom: 2em;
}

.timer, .moves {
  margin: 0.25em 0 0 2em;
}

.game-container {
  box-sizing: border-box;
  margin-left: 4em;
}

.game-cubes {
  background-color: #1a1111;
  border: 10px black solid;
  display: table-cell;
  height: 8em;
  width: 8em;
}

.active {
  background-color: #78FFD6;
}
