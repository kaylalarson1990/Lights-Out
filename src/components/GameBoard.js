import React from 'react';
import '../GameBoard.css';
import StopWatch from './StopWatch';
import Grid from './Grid';
import _ from 'lodash';

const indices = [0, 1, 2, 3, 4];

export default class GameBoard extends React.Component {
  constructor(props) {
    super(props);
    this.state = { size: 5, data: this.randomizeGameBoard(5), moves: 0 }
  }

  restartGame = () => {
    window.location.reload(false);
  }

  createGameBoard = (size, toggle) => {
    const board = [];
    _.forEach(indices, (row) => {
      board.push([]);
      _.forEach(indices, (column) => board[row].push(toggle(board, row, column)))
    });
    return board;
  }

  randomizeGameBoard = (size) => {
    return this.createGameBoard(size, () => Math.random() > 0.5);
  }

  targetAdjacent = (row1, column1, row2, column2) => {
    const rows = Math.abs(row1 - row2);
    const columns = Math.abs(column1 - column2);
    return (
      (rows == 0 && columns == 0) || (rows == 1 && columns == 0) || (rows == 0 && columns == 1)
    )
  }

  toggleLight = (row, column) => {
    const { data, moves } = this.state;
    const newData = this.createGameBoard(this.state.size, (board, r, c) => {
      return data[r][c] ^ this.targetAdjacent(row, column, r, c);
    });
    this.setState({ data: newData, moves: moves + 1 });
  }

  render() {
    const gameCubes = [];
    const { data, moves } = this.state;

    _.forEach(indices, (row) => {
      gameCubes.push(<Grid lights={ data[row] } row={ row } click={ this.toggleLight } />);
    });

    return (
      <div>
        <div className="header">
          <h1 className="title">Lights Out</h1>
          <div className="game-options">
            <button className="restart" onClick={ this.restartGame }>Restart</button>
            <StopWatch />
            <p className="moves">Moves: { moves }</p>
          </div>
        </div>
        <div>
          { gameCubes }
        </div>
      </div>
    );
  }
};
