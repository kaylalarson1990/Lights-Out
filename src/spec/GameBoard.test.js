import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import GameBoard from '../components/GameBoard';

test('renders Lights Out title', () => {
  const { getByText } = render(<GameBoard />);
  const linkElement = getByText(/Lights Out/i);
  expect(linkElement).toBeInTheDocument();
});

const title = "Lights Out";
let wrapper = shallow(<GameBoard>{title}</GameBoard>);

describe('GameBoard', () => {
  it('should render the GameBoard Component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the GameBoards children', () => {
    expect(wrapper.find('h1').text()).toEqual(title);
    expect(wrapper.find('td').exists()).toBeTruthy();
  });

  it('toggles class active on click', () => {
    const gridCube = wrapper.find('td').first();
    const activate = gridCube.simulate('click');
    expect(activate).toMatchSnapshot();
    expect(wrapper.state('active')).toEqual(true);
  });
});

describe('reloads page on game start', () => {
  const { reload } = window.location;
  const reloadFn = () => {
    window.location.reload(true);
  }
  beforeAll(() => {
    Object.defineProperty(window.location, 'reload', {
      configurable: true,
    });
    window.location.reload = jest.fn();
  });

  afterAll(() => {
    window.location.reload = reload;
  });

  it('mocks reload function', () => {
    expect(jest.isMockFunction(window.location.reload)).toBe(true);
  });

  it('calls reload function', () => {
    reloadFn();
    expect(window.location.reload).toHaveBeenCalled();
  });
});
