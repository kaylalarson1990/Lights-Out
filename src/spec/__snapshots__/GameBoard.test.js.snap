// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard should render the GameBoard Component correctly 1`] = `
<Fragment>
  <div
    className="header"
  >
    <h1
      id="title"
    >
      Lights Out
    </h1>
    <div
      className="sub-header"
    >
      <button
        onClick={[Function]}
      >
        Restart
      </button>
      <p
        id="timer"
      >
        Timer:
      </p>
    </div>
  </div>
  <div
    className="game-container"
  >
    <table>
      <thead>
        <tr>
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
        </tr>
      </thead>
    </table>
    <table>
      <thead>
        <tr>
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
        </tr>
      </thead>
    </table>
    <table>
      <thead>
        <tr>
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
        </tr>
      </thead>
    </table>
    <table>
      <thead>
        <tr>
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
        </tr>
      </thead>
    </table>
    <table>
      <thead>
        <tr>
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
          <td
            onClick={[Function]}
          />
        </tr>
      </thead>
    </table>
  </div>
</Fragment>
`;

exports[`GameBoard toggles class active on click 1`] = `
<td
  onClick={[Function]}
/>
`;
